File: Coursera-Machine-Learning/Matlab-Code/machine-learning-ex3/ex3/predict.m
function p = predict(Theta1, Theta2, X)
%PREDICT Predict the label of an input given a trained neural network
%   p = PREDICT(Theta1, Theta2, X) outputs the predicted label of X given the
%   trained weights of a neural network (Theta1, Theta2)

% Useful values
m = size(X, 1);
num_labels = size(Theta2, 1);

% You need to return the following variables correctly 
p = zeros(size(X, 1), 1);

% ====================== YOUR CODE HERE ======================
% Instructions: Complete the following code to make predictions using
%               your learned neural network. You should set p to a 
%               vector containing labels between 1 to num_labels.
%
% Hint: The max function might come in useful. In particular, the max
%       function can also return the index of the max element, for more
%       information see 'help max'. If your examples are in rows, then, you
%       can use max(A, [], 2) to obtain the max for each row.
%

% Add bias node
X = [ones(size(X, 1), 1) X];

for j = 1:m
  % Get training example at j-th of X dataset
  x = X(j, :);
  % compute z2
  a1 = x';
  z2 = Theta1 * a1;
  % compute a2
  a2 = sigmoid(z2);
  
  % compute z3
  a2 = [1; a2];
  z3 = Theta2 * a2;
  % compute a3
  a3 = sigmoid(z3);
  
  % compute output
  h = a3;
  
  % Get element and its index has highest probability in output vector above
  [max_value max_index] = max(h)
  
  % convert to p
  p(j) = max_index;
  
endfor

return




% =========================================================================


end

File: Coursera-Machine-Learning/Python-Code/ex3/predict.py
import numpy as np
from sigmoid import sigmoid

def predict(Theta1, Theta2, X):
    # Pre-processing for computing in numpy.
    if X.ndim == 1:
        X = X.reshape(1, len(X)) # (1, len(X))   
    
    m = X.shape[0] # num of training examples
    num_labels = Theta2.shape[0]
    
    p = np.zeros((m, 1))
    
    print('X.shape  = {0}'.format(X.shape))
    # Add bias
    X = np.hstack((np.ones((m, 1)), X))
    
    for j in range(m):
        # Get training examples at j-th of X dataset
        x = X[j, :] # (401, )
        x = x.reshape((X.shape[1], 1)) # (401, 1)
        # Compute z2
        a1 = x # (401, 1)
        z2 = Theta1.dot(a1) # Theta1: (25, 401) --> (25, 1)
        # Compute a2
        a2 = sigmoid(z2) # --> (25, 1)
        
        # Compute z3
        a2 = np.vstack((1, a2)) # --> (26, 1)
        z3 = Theta2.dot(a2)
        
        # Compute a3
        a3 = sigmoid(z3)
        
        # Compute output
        h = a3
        
        # Get element and its index has highest prob in output vector above
        max_index = np.argmax(h) # 0 .. 9
        
        # Convert to output p
        p[j] = max_index + 1

        '''
        # Fix zero index for predicting
        if max_index > 0:
            p[j] = 
        else:
            p[j] = max_index
        '''
    return p